token-contracts/OToken/src/main/java/finance/omm/score/tokens/OTokenImpl.java

```diff
@@ -0,0 +1,550 @@
+            Context.revert(TAG +":  Transfer error:Transfer cannot be allowed");
+        }
+
+        Map<String, BigInteger> previousBalances = this.executeTransfer(from, to, value);
+        this.balances.set(from, balanceFrom.subtract(value) );
+        this.balances.set(to, this.balances.getOrDefault(to, ZERO).add(value) );
+        this.callRewards(previousBalances.get("fromPreviousPrincipalBalance"),
+                previousBalances.get("toPreviousPrincipalBalance"), previousBalances.get("beforeTotalSupply"),
+                from,
+                to);
+
+        if (to.isContract()) {
+            /*
+            If the recipient is SCORE,
+            then calls `tokenFallback` to hand over control.
+            */
+            Context.call(to, "tokenFallback", from, value, data);
+        }
+
+        // Emits an event log `Transfer`
+        this.Transfer(from, to, value, data);
+    }
+
+    /**
+    Creates amount number of tokens, and assigns to account
+    Increases the balance of that account and total supply.
+    This is an internal function.
+    @param account: The account at which token is to be created.
+    @param amount: Number of tokens to be created at the `account`.
+    */
+    protected void mint(Address account, BigInteger amount) {
+
+        if (amount.compareTo(ZERO) < 0) {
+            Context.revert(TAG + ": Invalid value: " + amount + " to mint");
+        }
+
+        this.totalSupply.set(this.totalSupply.get().add(amount) );
+        this.balances.set(account, this.balances.getOrDefault(account, ZERO).add(amount));
+
+        // Emits an event log Mint
+        this.Transfer(AddressConstant.ZERO_ADDRESS, account, amount, "mint".getBytes());
+    }
+
+    /**
+    Destroys `amount` number of tokens from `account`
+    Decreases the balance of that `account` and total supply.
+    This is an internal function.
+    @param account: The account at which token is to be destroyed.
+    @param amount: The `amount` of tokens of `account` to be destroyed.
+    */
+    protected void burn(Address account, BigInteger amount) {
+        if (amount.equals(ZERO)) {
+            return;
+        }
+
+        if (amount.compareTo(ZERO) < 0) {
+            Context.revert( TAG + ": Invalid value: "+ amount + " to burn");
+        }
+        BigInteger actualTotalSupply = this.totalSupply.get();
+        BigInteger userBalance = this.balances.getOrDefault(account, ZERO);
+        if (amount.compareTo(actualTotalSupply) > 0) {
+            Context.revert( TAG + ": " + amount+ " is greater than total supply :" +actualTotalSupply );
+        }
+        if (amount.compareTo(userBalance) > 0) {
+            Context.revert(TAG + ": Cannot burn more than user balance. Amount to burn: " + amount + ", User Balance:" + userBalance);
+        }
+
+        this.totalSupply.set(actualTotalSupply.subtract(amount));
+        this.balances.set(account, this.balances.getOrDefault(account, ZERO).subtract(amount));
```

Comment:
userBalance has been set already. it can be used instead of getting variable again

Reply:
done